plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/procedure/MigrateProcedure.java

```diff
@@ -0,0 +1,354 @@
+    {
+        Map<String, String> icebergTableProperties = new HashMap<>();
+
+        // Copy all Hive table properties so that we can undo the migration easily. This is same as Spark implementation.
+        icebergTableProperties.putAll(hiveTableProperties);
+        EXCLUDED_PROPERTIES.forEach(icebergTableProperties::remove);
+
+        icebergTableProperties.put("migrated", "true");
+        icebergTableProperties.putIfAbsent("location", location);
+        icebergTableProperties.put(PROVIDER_PROPERTY_KEY, PROVIDER_PROPERTY_VALUE);
+        icebergTableProperties.put(METADATA_LOCATION_PROP, location);
+        icebergTableProperties.put(DEFAULT_NAME_MAPPING, toJson(nameMapping));
+        icebergTableProperties.put(FORMAT_VERSION, String.valueOf(formatVersion));
+
+        return ImmutableMap.copyOf(icebergTableProperties);
+    }
+
+    private Schema toIcebergSchema(List<Column> columns)
+    {
+        List<Types.NestedField> icebergColumns = new ArrayList<>();
+        for (Column column : columns) {
+            int index = icebergColumns.size();
+            org.apache.iceberg.types.Type type = toIcebergType(typeManager.getType(column.getType().getTypeSignature()));
+            Types.NestedField field = Types.NestedField.of(index, false, column.getName(), type, column.getComment().orElse(null));
+            icebergColumns.add(field);
+        }
+        org.apache.iceberg.types.Type icebergSchema = Types.StructType.of(icebergColumns);
+        AtomicInteger nextFieldId = new AtomicInteger(1);
+        icebergSchema = TypeUtil.assignFreshIds(icebergSchema, nextFieldId::getAndIncrement);
+        return new Schema(icebergSchema.asStructType().fields());
+    }
+
+    public Map<String, Optional<Partition>> listAllPartitions(HiveMetastore metastore, io.trino.plugin.hive.metastore.Table table)
+    {
+        List<String> partitionNames = table.getPartitionColumns().stream().map(Column::getName).collect(toImmutableList());
+        Optional<List<String>> partitions = metastore.getPartitionNamesByFilter(table.getDatabaseName(), table.getTableName(), partitionNames, TupleDomain.all());
+        if (partitions.isEmpty()) {
+            return ImmutableMap.of();
+        }
+        return metastore.getPartitionsByNames(table, partitions.get());
+    }
+
+    private List<DataFile> buildDataFiles(ConnectorSession session, HiveStorageFormat format, String location, PartitionSpec partitionSpec, StructLike partition, NameMapping nameMapping)
+            throws IOException
+    {
+        // TODO: Introduce parallelism
+        TrinoFileSystem fileSystem = fileSystemFactory.create(session);
+        FileIterator files = fileSystem.listFiles(location);
+        ImmutableList.Builder<DataFile> dataFiles = ImmutableList.builder();
+        while (files.hasNext()) {
+            FileEntry file = files.next();
+            if (!file.path().contains("/_") && !file.path().contains("/.")) {
```

Comment:
The path is the full URI, so we could skip directories that start with underscores. This probably should do
```java
String name = entry.path().substring(entry.path().lastIndexOf('/') + 1);
if (!name.startsWith("_") && !name.startsWith(".")) {

Reply:
The suggested logic works when `file` is a directory, but the object can be a file path. e.g.
```
.../iceberg_data/tpch/test_migrate/.test_directory/foo.orc
```

The intention of the current logic is handling the both files and directories.